Sources/SecureXPC/Server/XPCAnonymousServer.swift

```diff
@@ -0,0 +1,44 @@
+//
+//  XPCAnonymousServer.swift
+//  
+//
+//  Created by Alexander Momchilov on 2021-11-28.
+//
+
+import Foundation
+
+internal class XPCAnonymousServer: XPCServer {
+    private let anonymousListenerConnection: xpc_connection_t
+
+    internal override init() {
+        self.anonymousListenerConnection = xpc_connection_create(nil, nil)
+        super.init()
+
+        // Start listener for the new anonymous connection, all received events should be for incoming client connections
```

Comment:
Why is this happening in the initializer instead of in `start()`? Having it occur as part of the `start()` function would make it consistent with `XPCMachServer` and `XPCServiceServer`.

Reply:
We can’t use `start()` because it returns `Never`, whereas this server doesn’t need to take over the main thread, or anything like that.

Reply:
I'm not sure I follow about what's wrong with using `start()`? It never returns by choice, not because it must. For example `XPCMachServer` doesn't have to take over the main thread either, it's just convenient for it to do so due to how most helper tools are likely to be written. The onus certainly could be put on the API caller to handle keeping the main thread alive.

This inconsistent behavior is definitely undesirable particularly because only `XPCServer` is the exposed type. That sometimes calling `start()` on an `XPCServer` results in a `fatalError` is too unpredictable. I can absolutely imagine API users relying having functionality that operates on a `XPCServer` where the underlying server type may differ at runtime (I'm actually considering doing this so that I have shared code which can run as either an XPC Service or as a Launch Agent using Mach services).

If there's a good reason to have two different ways to start a server then I think a reasonable approach would be either to create two different start functions or add a `Bool` parameter to specify whether to dispatch the main thread as part of starting. The disadvantage of the latter is its return type can't be `Never` then even if it doesn't actually return.

Reply:
> I'm not sure I follow about what's wrong with using start()?

See [the usage in the `RoundTripIntegrationTest`](https://github.com/trilemma-dev/SecureXPC/pull/25/commits/dd05992d541827a34e122fd96e4cb5238d1d4bca#diff-1d0119acaf113027616db5ae088a57770b607260d6911dc23b22d29fd2c82c54R15). If we called a `start() -> Never` method, the main thread would block, and it would be game over.

> The onus certainly could be put on the API caller to handle keeping the main thread alive.

What do you mean by "keeping the main thread alive"?

> This inconsistent behavior is definitely undesirable particularly because only `XPCServer` is the exposed type. That sometimes calling `start()` on an `XPCServer` results in a `fatalError` is too unpredictable. 

You're right, that's pretty jank.

> I'm actually considering doing this so that I have shared code which can run as either an XPC Service or as a Launch Agent using Mach services

Funny enough, I was building that too! I wanted to have an easier debugging experience for most things that fully need to run as root.

> If there's a good reason to have two different ways to start a server then I think a reasonable approach would be either to create two different start functions or add a Bool parameter to specify whether to dispatch the main thread as part of starting. The disadvantage of the latter is its return type can't be Never then even if it doesn't actually return.

This design is definitely under-baked. I'll look into it.

Reply:
> What do you mean by "handle keeping the main thread alive"?

I mean have the caller invoke `dispatchMain()` (or equivalent) themselves.

Reply:
I'm rebased on top of the new `NonBlockingStartable` changes. How does it look now?

How do you think `startAndBlock()` should work for `XPCAnonymousServer`? Arbitrarily calling `dispatchMain` or `RunLoop.main.run()` doesn't really "fit"

Reply:
I think ideally we'd want to do is block the calling thread, not the main thread (unless that happens to be the calling thread, do you know if there's a robust way to do that? (This makes me also wonder what happens when `dispatchMain()` is called from not the main thread, does it genuinely not return *and* block the main thread?)

If that's approach isn't feasible. We can remove `startAndBlock()` from `XPCServer` and then introduce a `BlockingStartable` interface as well. I rather dislike this approach though as it makes the common case for the non-anonymous servers (which are far more important to the API) a bit more complex to understand.

Reply:
(This makes me also wonder what happens when `dispatchMain()` is called from not the main thread, does it genuinely not return and block the main thread?)

You just get a crash:

```
Crashed Thread:        9

Exception Type:        EXC_BAD_INSTRUCTION (SIGILL)
Exception Codes:       0x0000000000000001, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Termination Reason:    Namespace SIGNAL, Code 4 Illegal instruction: 4
Terminating Process:   exc handler [6395]

Application Specific Information:
BUG IN CLIENT OF LIBDISPATCH: dispatch_main() must be called on the main thread
Performing @selector(echo:) from sender NSButton 0x7f8caf7699d0
```

> I think ideally what we'd want to do is block the calling thread

This would fit the "AndBlock" semantics of the method, but ultimately, it's useless. There's almost never a good reason to be blocking a thread like that.

Reply:
> You just get a crash

Ah, yeah we should document this. I'll add an issue to do so.

---

On thinking about this more, it seems actually having `startAndBlock()` for anonymous server call `dispatchMain()` would be useful for testing purposes. By testing I don't mean _us_ writing integration tests for SecureXPC, but instead for users of our API to be able to sub out one of the other server types and check that their code actually works the same as it would when running in an external process. Am I wrong in thinking that's useful?